runtime-modules/storage/src/lib.rs

```diff
@@ -4459,97 +4444,110 @@ impl<T: Trait> Module<T> {
         Ok(())
     }
 
-    fn construct_objects_to_upload(
-        op: &BagOperation<T>,
-    ) -> Result<Vec<DataObject<BalanceOf<T>>>, DispatchError> {
+    fn construct_objects_to_upload(op: &BagOperation<T>) -> DataObjAndDeletionPrizeAndObjSize<T> {
         match &op.params {
-            BagOperationParams::<T>::Delete => Ok(Default::default()),
-            BagOperationParams::<T>::Create(_, list, _, _) => {
-                Self::construct_objects_from_list(list)
+            BagOperationParams::<T>::Delete => Ok((Default::default(), Zero::zero(), Zero::zero())),
+            BagOperationParams::<T>::Create(bag_deletion_prize, list, _, _) => {
+                Self::construct_objects_from_list(list, *bag_deletion_prize)
+            }
+            BagOperationParams::<T>::Update(list, _) => {
+                Self::construct_objects_from_list(list, Zero::zero())
             }
-            BagOperationParams::<T>::Update(list, _) => Self::construct_objects_from_list(list),
         }
     }
 
     fn construct_objects_from_list(
         list: &[DataObjectCreationParameters],
-    ) -> Result<Vec<DataObject<BalanceOf<T>>>, DispatchError> {
+        bag_deletion_prize: BalanceOf<T>,
+    ) -> DataObjAndDeletionPrizeAndObjSize<T> {
+        let deletion_prize = Self::data_object_deletion_prize_value();
         list.iter()
             .map(|param| {
                 Self::upload_data_objects_checks(param).and({
                     Ok(DataObject {
                         accepted: false,
-                        deletion_prize: Self::data_object_deletion_prize_value(),
+                        deletion_prize,
                         size: param.size,
                         ipfs_content_id: param.ipfs_content_id.clone(),
                     })
                 })
             })
-            .collect()
+            .try_fold(
+                Ok((Vec::new(), bag_deletion_prize, 0)),
+                |acc: DataObjAndDeletionPrizeAndObjSize<T>, obj| {
+                    obj.map(|obj| {
+                        acc.map(|(mut acc_obj, acc_deletion_prize_request, acc_size)| {
+                            let acc_deletion_prize_request =
+                                acc_deletion_prize_request.saturating_add(obj.deletion_prize);
+
+                            let acc_size = acc_size.saturating_add(obj.size);
+
+                            acc_obj.push(obj);
+
+                            (acc_obj, acc_deletion_prize_request, acc_size)
+                        })
+                    })
+                },
+            )
+            .and_then(|x| x)
     }
 
     fn construct_objects_to_remove(
```

Comment:
This function will probably change in a future refactoring since it has became quite complex.
For now please just add a doc line like describing the main units of work done

Reply:
Now these functions have a comment explaining the workflow better.
I added a function to simplify, 
```rust
    //Sums the accumulated prize and size to the new object in the iteration.
    fn calculate_acc_size_and_acc_prize(
        mut acc_obj: Vec<DataObject<BalanceOf<T>>>,
        acc_deletion_prize: BalanceOf<T>,
        acc_size: u64,
        obj: DataObject<BalanceOf<T>>,
    ) -> (Vec<DataObject<BalanceOf<T>>>, BalanceOf<T>, u64) {
        let acc_deletion_prize: T::Balance = acc_deletion_prize.saturating_add(obj.deletion_prize);

        let acc_size: u64 = acc_size.saturating_add(obj.size);

        acc_obj.push(obj);

        (acc_obj, acc_deletion_prize, acc_size)
    }
```
```rust
.try_fold(
        //As an initial value we set the fold with an empty vec, the bag deletion prize,
        //(it's zero in case of an update), and 0 size.
        Ok((Vec::new(), bag_deletion_prize, 0)),
        |acc: DataObjAndDeletionPrizeAndObjSize<T>, obj| {
               obj.map(|obj| {
                        acc.map(|(acc_obj, acc_deletion_prize_request, acc_size)| {
                            Self::calculate_acc_size_and_acc_prize( <-----------------------------------------------
                                acc_obj,
                                acc_deletion_prize_request,
                                acc_size,
                                obj,
                            )
                        })
                    })
                },
            )
       //Removes outer result
       //(Result<Result[Item1 + Item2 + Item3 + ...]>) -> Result[Item1 + Item2 + Item3 + ...]
       .and_then(|x| x)
```